Manual chapters on parallelism and memory model

kayceesrk opened the pull request:
This PR adds a manual chapter on parallelism features in OCaml. ~Many sections are yet to be written, but I thought it might be a good idea to open the PR to get some early feedback.~

The rendered HTML pages are at:
* Parallelism -- https://kcsrk.info/webman/manual/parallelism.html
* Memory model: the hard bits -- https://kcsrk.info/webman/manual/memorymodel.html

- [x] (Section) Memory model: the easy bits. The aim is to cover only DRF-SC.
- [x] (Chapter) Memory model: the hard bits. Cover non-SC behaviours and features that do not respect the memory model.
- [x] (Section) Atomics
- [x] (Section) Blocking synchronization
- [x] Update the chapter on [the threads library](https://v2.ocaml.org/manual/libthreads.html) and its interaction with domains.

kayceesrk committed:
Add manual chapters on parallelism and memory model

kayceesrk commented:
I have now included everything that I wanted to say in the parallelism chapter. CC @gasche @fpottier.

OlivierNicole commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,666 @@
+The OCaml standard library exposes low-level primitives for parallel
+programming. We recommend the users to utilise higher-level parallel
+programming libraries such as
+\href{https://github.com/ocaml-multicore/domainslib}{domainslib}. This
+tutorial will first cover the high-level parallel programming using domainslib
+followed by low-level primitives exposed by the compiler. This tutorial will
+first cover the parallelism facilities provided by the standard library
+followed by high-level parallel programming using domainslib.
+
```
Repeated sentence (with the order of actions reversed)

OlivierNicole commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,666 @@
+
+\subsection{s:join}{Joining domains}
+
+We shall use the recursive program to compute the nth Fibonacci number as a
```
the recursive -> this recursive

OlivierNicole commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,666 @@
+  Range (min … max):    1.181 s …  1.202 s    10 runs
+\end{verbatim}
+
+We see that it takes around 1.2 seconds to compute the nth Fibonacci number.
```
to compute the **42nd** Fibonacci number (in this case)

avsm commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,666 @@
+The OCaml standard library exposes low-level primitives for parallel
+programming. We recommend the users to utilise higher-level parallel
+programming libraries such as
+\href{https://github.com/ocaml-multicore/domainslib}{domainslib}. This
+tutorial will first cover the high-level parallel programming using domainslib
+followed by low-level primitives exposed by the compiler. This tutorial will
+first cover the parallelism facilities provided by the standard library
```
Looks like a repeated sentence here (one of which is contradictory)

OlivierNicole reviewed (commented):
I just found a few typos, otherwise I have no further comment about the contents, it looks good to me.

fpottier commented:
@kayceesrk Could you update the [rendered version](https://kcsrk.info/webman/manual/parallelism.html)? Thanks!

kayceesrk commented:
@fpottier 

I've pushed the rendered page. It should go live in a few minutes. Thanks for taking a look.

avsm commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,690 @@
+followed by low-level primitives exposed by the compiler.
+
+OCaml distinguishes concurrency and parallelism and provides distinct
+mechanisms for expressing them. Concurrency is overlapped execution of tasks
```
Can you `\ref` into the main manual here? Section `s:effects-concurrency` has a nice definition of exactly what a task is and the states it can be in.

kayceesrk replied:
Done.

OlivierNicole commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,238 @@
+other cores. Since "a" and "b" are different memory locations, the reads of "a"
+and "b" may both witness the initial values, leading to the assertion failure.
+This behaviour can be explained if a load is allowed to be reordered before a
+preceding store to different memory location.
```
```suggestion
This behaviour can be explained if a load is allowed to be reordered before a
preceding store to a different memory location.
```

kayceesrk replied:
Done.

avsm commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,691 @@
+structures for allocating memory. Hence, they are relatively expensive to
+create and tear down.
+
+\emph{\textbf{It is recommended that the users spawn exactly one domain per
```
Not "exactly one domain per available core" -- it's ok to have fewer domains than available cores too :-) 

"It is recommended that programs do not spawn a number of domains that exceed the number of available cores"?

kayceesrk commented:
Thanks for the reviews so far. 

For reviewers who have write permission to `ocaml/ocaml`, please feel free to directly push minor edits to `kayceesrk:domains_manual` branch.

fabbing commented:
Chapter 9 looks good to me!

kayceesrk commented:
Thanks @fabbing for having a look.

avsm commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,695 @@
+
+OCaml distinguishes concurrency and parallelism and provides distinct
+mechanisms for expressing them. Concurrency is overlapped execution of
+tasks~\ref{s:effects-concurrency} whereas parallelism is simultaneous execution
```
this needs to be` (\S\ref{s:effect-concurrency})` or similar, or it'll just be a bare number when rendered.

kayceesrk replied:
Fixed.

xavierleroy commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,496 @@
+\href{https://doi.org/10.1145/3192366.3192421}{Bounding Data Races in Space and
+Time}.
+
+\section{s:why_weak_memory}{Why do we need a weak memory model?}
```
Nitpicking here: the memory model is not weak, it's the consistency guarantees that are weak.  (Maybe "relaxed" is preferable to "weak"; not sure.)  So what about "Why weakly-consistent memory?" or "Relaxed consistency in the memory model" or something like this?

xavierleroy commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,695 @@
+control to each other. While concurrency is a program structuring mechanism,
+parallelism is a mechanism to make your programs run faster. If you are
+interested in the concurrent programming mechanisms in OCaml, please refer to
+the section \ref{s:effect-handlers} on effect handlers.
```
and perhaps to the Threads library?  Not sure.

xavierleroy commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,695 @@
+may or may not overlap in time. Tasks may execute concurrently by yielding
+control to each other. While concurrency is a program structuring mechanism,
+parallelism is a mechanism to make your programs run faster. If you are
+interested in the concurrent programming mechanisms in OCaml, please refer to
```
"in concurrent programming in OCaml" or "in the concurrent programming mechanisms of OCaml".

xavierleroy commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,695 @@
+
+The atomic variables can be used for low-level synchronization between the
+domains. The following example uses an atomic variable to exchange a message
+between two domains.
```
… but busy-waits like crazy.  I think users should not be encouraged to write busy-waiting loops, at least not before they have been exposed to condition variables.

kayceesrk replied:
I've switched the order of the blocking synchronisation and atomics sections.

xavierleroy commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,695 @@
+stack.
+
+\section{s:par_sync}{Blocking synchronisation}
+
```
I wonder whether this section should come before the section on atomics.  Lock-free algorithms are more advanced than algorithms using locks.

kayceesrk replied:
I've switched the order of the blocking synchronisation and atomics sections.

xavierleroy commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,695 @@
+modules. These modules are the same as those used to synchronise threads
+created by the threads library~\ref{c:threads}. For clarity, in the rest of
+this chapter, we shall call the threads created by the threads library as
+systhreads. The following program implements a blocking version of a concurrent
```
I'm not a fan of the "systhreads" terminology, but don't have a better suggestion yet.

kayceesrk replied:
Not a fan either. But I wanted to call this something other than "threads" to avoid confusion. Let me know if you have a better name.

xavierleroy commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,695 @@
+  type 'a t = {
+    mutable contents: 'a list;
+    mutex : Mutex.t;
+    condition : Condition.t
```
What about a more descriptive name for this condition variable?
```suggestion
    nonempty : Condition.t
```
Typically, there's only one mutex but several condition variables associated with a data structure, so different, descriptive names for the condvars are needed.

xavierleroy reviewed (commented):
Nice work, thanks!  Some minor remarks below.

kayceesrk commented:
The memory model chapter is also complete now. 

I've included the operational view of the memory model in the memory model chapter. Given that the OCaml memory model has a straightforward small-step operational semantics, it is useful to include it in the manual for our expert users rather than only have it in the [PLDI 2018 paper (section 3)](https://kcsrk.info/papers/pldi18-memory.pdf).

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+can be explained through some interleaving of the operations from different
+domains in the program.
+
+For example, consider the program with two domains "d1" and "d2" executing in
```
This example is not very clear to me. We have two functions, both with local variables called `b`. Presumably these are going to be called with the same ref cell, but this isn't stated anywhere. The example around line 100 avoids this problem and shows the complete program. It might be better to start in that style.

maranget replied:
I'd also favor swapping  the examples. For an additional reason: blame relaxation of SC on hardware optimisations :)

kayceesrk replied:
Reworked the section for clarity.

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+memory model. While programming directly under the relaxed memory model is
+difficult, the memory model also describes the conditions under which a program
+will only exhibit sequentially consistent behaviours. This guarantee is known
+as data race freedom implies sequential consistency (DRF-SC). In this section,
```
I think it's a little easier to read as:
```suggestion
as \emph{data race freedom implies sequential consistency} (DRF-SC). In this section,
```

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+\emph{(TODO: Include semaphores and condition variables in inter-domain
+actions.)}
+
+On the other hand, an intra-domain actions can neither be observed nor
```
```suggestion
On the other hand, intra-domain actions can neither be observed nor
```

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+
+A totally ordered list of actions executed by the abstract machine is called
+the \emph{execution trace}. There might be several execution traces for a given
+program due to non-determinism.
```
```suggestion
A totally ordered list of actions executed by the abstract machine is called
an \emph{execution trace}. There might be several possible execution traces for a given
program due to non-determinism.
```

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+
+DRF-SC is a strong guarantee for the programmers. Programmers can use
+sequential reasoning (executing one inter-domain action to identify whether
+their program has a data race. In particular, they do not need to reason about
```
This looks like a copy-and-paste error. There's a missing `)` somewhere, anyway.

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+program is data race free is made, they do not need to worry about reorderings
+in their code.
+
+For programs with data races, OCaml memory model ensures that there are no
```
```suggestion
For programs with data races, OCaml's memory model ensures that there are no
```
or
```suggestion
For programs with data races, the OCaml memory model ensures that there are no
```
(occurs in several places)

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+    \emph{program order}. If an action "x" precedes another action "y" in
+    program order, then "x" precedes "y" in happens-before order.
+  \item If "x" is an initialising write to a memory location (atomic or
+    otherwise) and "y" is any read or write to that memory location, then "x"
```
```suggestion
    otherwise) and "y" is any other read or write to that memory location, then "x"
```

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+\begin{caml_example*}{verbatim}
+(* Has data race *)
+let r = ref 0
+let d1 () = ref 1
```
Is this a typo? I assume `d1` is supposed to be accessing `r` somehow.

kayceesrk replied:
Should be 

```ocaml
let d1 () = r := 1
```

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+
+Each domain is equipped with a \emph{frontier}, which is a map from non-atomic
+locations to timestamps. Intuitively, each domain's frontier records, for each
+non-atomic locations, the latest write known to the thread. More recent writes
```
```suggestion
non-atomic location, the latest write known to the thread. More recent writes
```

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+
+Suppose domain "d2" writes the value "10" to "c" ("c := 10"). We pick a new
+timestamp "t9" for this write such that it is later than "d2"'s frontier at
+"c". Note a subtlety here: this new timestamp may not be later than everything
```
```suggestion
"c". Note a subtlety here: this new timestamp might not be later than everything
```
("may not" is unclear as it can mean either "must not" or "might not")

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+
+During atomic reads, the frontier of the location is merged into that of the
+domain performing the read. For example, suppose "d1" reads "B". The read
+returns "5", and "d1"'s frontier merged with "B"'s frontier, choosing the later
```
This sounds like the merged frontier is returned as the result of the read. Perhaps it should say that d1's frontier is updated from B's?

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+\end{verbatim}
+
+During atomic writes, the value held by the atomic location is updated. The
+frontiers of both the reading domain and that of the location being read is
```
```suggestion
frontiers of both the reading domain and that of the location being read are
```

talex5 commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+in their code.
+
+For programs with data races, OCaml memory model ensures that there are no
+out-of-thin-air (OOTA) values. Reads always return a value written to a memory
+location (including possibly be the initial write). Data races do not cause
```
```suggestion
out-of-thin-air (OOTA) values. Reads always return a value written to that memory
location. Data races do not cause
```
(I assume "including possibly be the initial write" means "an initial write is a write", but that seems obvious from the name)

maranget replied:
Notice that even in the archetypal example of OOTA, every read returns a value written by some write:
```
Initially *x=0, *y=0

r0 = *x  | r1 = *y
*y = r0  | *x = r1
exists r0 == 42 && r1 == 42
```
For instance the read  `r0 = *x` reads from write `*x = r1` .

Anyway, defining OOTA is tricky (is there a definition beyond examples?).

kayceesrk replied:
We don't have a definition of OOTA. Given that the OCaml memory model does not exhibit OOTA, it may be better to drop the reference to OOTA from the manual.

talex5 reviewed (commented):
Note: I only looked at the memorymodel file here.

Perhaps it's obvious to most readers, but it might be useful to clarify that an OCaml location is not the same as a physical memory location. e.g. the location returned by `ref` is a new location, even though it may reuse a physical address that was previously GC'd.

maranget commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,695 @@
+Atomic ref count: 2000000
+\end{verbatim}
+
+As one can see, the final value of the non-atomic counter is lower than one
+might naively expect. In fact, this program exhibits a data race: there are
+unsynchronised concurrent accesses to the non-atomic counter.
```
Hi @kayceesrk. There is a difficulty here, as the output of the non-atomic counter is valid SC output. Hence, the example does  does not really illustrate the sentence just before:
> _One way to avoid data races is to use `Atomic` variables for inter-domain synchronisation. Importantly, unsynchronised concurrent accesses to a shared atomic location is not considered a data race. Hence, the program will only exhibit sequentially consistent behaviour._

In fact, the example more illustrates atomicity than SC. May I suggest introducing atomicity before the example. Then present the memory model properties of atomics. This would more or less amount to moving the quoted sentence downwards.

kayceesrk replied:
You are right. I've modified the text to talk about atomicity instead of data races.

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+other cores. Since "a" and "b" are different memory locations, the reads of "a"
+and "b" may both witness the initial values, leading to the assertion failure.
+This behaviour can be explained if a load is allowed to be reordered before a
+preceding store to a different memory location.
```
```suggestion
preceding store to a different memory location, which is the case in the presence of in-core store-buffers.
```

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+    operation on the mutex in the execution trace, then "x" precedes "y" in
+    happens-before order.
+\end{itemize}
+
```
Is this list complete? It looks like an extension of the [ocaml.cat](https://github.com/ocamllabs/ocaml-memory-model/blob/master/ocaml.cat) `hb` relation.

kayceesrk replied:
The list is missing happens-before ordering due to semaphores and condition variables. I can't think of anything else that is missing.

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+\subsection{s:datarace}{Data race}
+
+In a given trace, two actions are said to be \emph{conflicting} if they access
+the same non-atomic location, and at least one is a write.
```
Conflicting accesses originate from different concurrent agents (threads?)
```suggestion
the same non-atomic location, are performed by two different threads, and at least one is a write.
```

kayceesrk replied:
Not necessarily. Two accesses from the same thread may also be conflicting. But such a pair will not have a data race since the program order between the pair of accesses implies happens-before order. 

The definition of conflicting access closely follows the definition (9) from the PLDI paper on the OCaml memory model: https://kcsrk.info/papers/pldi18-memory.pdf. I've tried not to be creative here.

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+\begin{caml_example*}{verbatim}
+(* Has data race *)
+let r = ref 0
+let d1 () = ref 1
```
```suggestion
let d1 () = r := 1
```

kayceesrk replied:
Good catch. Thanks.

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+conflicting accesses, there is a data race.
+
+Both of the programs that we had seen in the section~\ref{s:why_relaxed_memory}
+have have data races, and hence, exhibit non sequentially consistent
+behaviours.
```
Typo (double "have") + data racy programs do not necessarily have non-SC behaviours, as illustrated by the simple example above. Tentative suggestion:
```suggestion
have data races. It is no surprise that they exhibit non sequentially consistent
behaviours.
```

kayceesrk replied:
Thanks.

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+\begin{caml_example*}{verbatim}
+(* No data race *)
+let a = [| 0; 1 |]
+let d1 = a.(0) <- 42
+let d2 = a.(1) <- 42
```
```suggestion
let d1 () = a.(0) <- 42
let d2 () = a.(1) <- 42
```

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+  mutable b : int
+}
+let r = {a = 0; b = 1}
+let d1 = r.a <- 42
+let d2 = r.b <- 42
```
```suggestion
let d1 () = r.a <- 42
let d2 () = r.b <- 42
```

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+
+OCaml memory model offers strong guarantees even for programs with data races.
+It offers what is known as \emph{local data race freedom sequential
+consistency(LDRF-SC)} guarantee. A formal definition of this property is beyond
```
Typo
```suggestion
consistency (LDRF-SC)} guarantee. A formal definition of this property is beyond
```

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+
+Since "s" and "o" are newly allocated in "d1", the only schedule where "d2"
+will read those objects is the one in which "d1" executes before "d2". Thanks
+to LDRF-SC, we can use sequential reasoning, which guarantees that "d2" will
```
I do not understand "_sequential reasoning_".

kayceesrk replied:
"Sequential reasoning" is defined when it is first introduced in the [subsection on DRF-SC](https://kcsrk.info/webman/manual/memorymodel.html#ss:drf_sc).

```
Programmers can use sequential reasoning i.e., reasoning by executing one 
inter-domain action after the other, ...
```

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+\end{verbatim}
+
+During atomic writes, the value held by the atomic location is updated. The
+frontiers of both the reading domain and that of the location being read is
```
This is a write? Tentative suggestion:
```suggestion
frontiers of both the writing domain and that of the location being written to is
```

kayceesrk replied:
That is correct. Thanks.

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+  assert (not (r1 = 0 && r2 = 0))
+\end{caml_example*}
+
+This program has a data race on "a" and "b", and hence, the program may observe
```
```suggestion
This program has a data race on "a" and "b", and hence, the program may exhibit
```
Not so sure given "observe" in the text that follows. However, I found bizarre having a program to "observe" without further precision.

kayceesrk replied:
"Exhibit" is indeed the right term. Thanks.

maranget commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,740 @@
+will read those objects is the one in which "d1" executes before "d2". Thanks
+to LDRF-SC, we can use sequential reasoning, which guarantees that "d2" will
+always see the initialising writes from "d1".
+
```
I found this section very interesting, thanks.

kayceesrk replied:
Thanks. Glad this is useful.

kayceesrk commented:
I believe I have addressed all the comments from @maranget and @talex5. Thanks!

shindere commented:
KC Sivaramakrishnan (2022/06/01 22:07 -0700):
> The memory model section is also complete now.
>
> I've included the operational view of the memory model in the memory
> model chapter. Given that the OCaml memory model has a straightforward
> small-step operational semantics, it is useful to include it in the
> manual for our expert users rather than only have it in the [PLDI 2018
> paper (section 3)](https://kcsrk.info/papers/pldi18-memory.pdf).

Will the manual contain a link to the paper, though?

kayceesrk commented:
It does already. The second paragraph of the memory model chapter says:

> This chapter is aimed at experts who would like to understand the details of
the OCaml memory model from a practitioner's standpoint. For a formal
definition of the OCaml memory model, its guarantees and the compilation to
hardware memory models, please have a look at the PLDI 2018 paper on
\href{https://doi.org/10.1145/3192366.3192421 }{Bounding Data Races in Space and
Time}.

shindere commented:
KC Sivaramakrishnan (2022/06/07 01:28 -0700):
> It does already. The second paragraph of the memory model chapter says:
>
> > This chapter is aimed at experts who would like to understand the details of
> the OCaml memory model from a practitioner's standpoint. For a formal
> definition of the OCaml memory model, its guarantees and the compilation to
> hardware memory models, please have a look at the PLDI 2018 paper on
> \href{https://doi.org/10.1145/3192366.3192421 }{Bounding Data Races in Space and
> Time}.

That's good! Thanks!

kayceesrk commented:
The PR has had a number of reviews but no approvals so far. I wonder when we are happy to merge the chapters :-) 

@fpottier are you planning to review the memory model chapter?

fpottier commented:
> @fpottier are you planning to review the memory model chapter?

I probably won't have time in the near future. Perhaps @jhjourdan might be able and willing to review it?

jhjourdan commented:
I will be at PLDI next week so I won't have time until the week after. If it's OK to you to wait 10-15 more days, then I could review this if needed.

fabbing commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,783 @@
+race.
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
```
Shouldn't this comment indicate there is a date race in this example?

kayceesrk replied:
Good spot. Fixed. Thanks.

kayceesrk commented:
@jhjourdan 

> I will be at PLDI next week so I won't have time until the week after. If it's OK to you to wait 10-15 more days, then I could review this if needed.

It would be useful to have your feedback. We can wait a bit longer for this. Thanks!

gadmm commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,783 @@
+
+\section{s:local_drf}{Local data race freedom}
+
+The OCaml memory model offers strong guarantees even for programs with data races.
```
Out of curiosity I searched for "safety" (as in memory safety) in this file and it returned nothing. The "easy bits" mention the absence of "crashes", so I wonder if it would not be a good idea to link up the informal claim with a more formal claim in a sentence or two in this section.

Also, constraints on the memory representation play a role, such as the tag being immutable from now on (and other things?). (This kind of details could be more important for people who are already interested by the memory representation, such as those doing things on the C side.)

kayceesrk replied:
> Out of curiosity I searched for "safety" (as in memory safety) in this file and it returned nothing. The "easy bits" mention the absence of "crashes", so I wonder if it would not be a good idea to link up the informal claim with a more formal claim in a sentence or two in this section.

Let me add a sentence about memory safety.

> Also, constraints on the memory representation play a role, such as the tag being immutable from now on (and other things?). (This kind of details could be more important for people who are already interested by the memory representation, such as those doing things on the C side.)

Would be good to include this in this page. Should we aim to do this as a follow-up PR? This PR is already large and has been up for a while now.

gadmm replied:
Sounds good. Note that there eventually ought to be a separate explanation of how to write C code that conforms to the OCaml memory model...

jhjourdan commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,732 @@
+
+\begin{verbatim}
+$ ocamlopt -o fib_twice.exe fib_twice.ml
+$ ./fib_twice.exe
```
```suggestion
$ ./fib_twice.exe 42
```

kayceesrk replied:
Done.

gadmm commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,783 @@
+let d1 () =
+  r := Some "Hello, world!"
+let d2 () =
+  match !r with
```
Is there is any guarantee that the load `!r` sees a recent value with this idiom? (e.g. in case of redundant load optimizations.)

kayceesrk replied:
Answered elsewhere, no.

gadmm commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,783 @@
+that when "d2" sees the "Some" variant when reading "r", the objects have been
+fully initialised?
+
+The answer is yes, and we can use LDRF to reason about this. Let us refactor the
```
Thanks, it was interesting to see LDRF applied here. But I think it conflates OCaml's notion of initializing writes with the notion from the PLDI paper. I think the discussion around dependency ordering was convincing-enough to assume a stronger result of LDRF in the manual than in the paper, but you could say more clearly that you do so.

kayceesrk replied:
Could you suggest the text that should go in here?

gadmm reviewed (commented):
The title of the section "LDRF > Message Passing", the one using non-atomics, makes think that it presents a way to publish data to multiple readers which is cheaper for the reader than using OCaml's atomics. After I asked about this at #10995, I think that there are still open questions about it.

1. Is there is any guarantee that the load `!r` sees a recent value, in the absence of any further synchronization between the two domains? Your reasoning goes “if `d2` sees `Some` then ...” but this leaves the door open for the function never seeing the updated value. Concretely, after inlining, I wonder if some optimization (e.g. RL in the PLDI paper) could have `d2` staying forever with the initial value `None` as a result of the load. (I have been thinking that a `Sys.opaque_identity` was missing in theory somewhere around the read in my example at #10995, without expectations that the memory model would justify this data-passing idiom.) One solution could be not to present it as a message-passing idiom but only as a memory-safety property. But I was wondering whether you have stronger guarantees in theory. (In practice, I know that all polling location are expected to possibly hide a write to `r`, so such optimizations are unlikely to happen.)
2. Previous point aside, it was interesting to see how LDRF could be used to reason about it. But as we have agreed at #10995, this usage follows from the OCaml memory model but not strictly from the PLDI paper, whose notion of initializing write is too weak even to prove publication safety with the stop-the-world GC. Stephen's explanations have been mostly convincing so far, but some work appears needed to even give it a proper definition. To be clear, I am fine with anticipating on results concerning the memory model implemented in practice in OCaml. However I have a nit and a comment:
    * The manual is not clear currently that the notion of initializing write is stronger than the one from the PLDI paper. The latter defines the initial write in such a way that it is determined too early to possibly create a well-formed OCaml value, whereas the OCaml memory model needs to leverage dependency ordering in a novel way (which is the key to making this message-passing idiom possible). I believe that the text should make a clearer distinction between the intended practical memory model for OCaml and what has been presented in the PLDI paper, indeed the reader might incorrectly infer that an argument for memory safety follows from the referenced paper.
    * Whether you intend to present it as a data-passing idiom or not, users might end up using it anyway in some use-cases, on the basis that it works in practice. From this point of view, it comes close to having a clearer and more useful guarantee where the reader does not merely see the initial (or later) write, but the values at least in their state when they were first published (think of data which is created by mutation and later published as immutable). (You have this guarantee I think with another `opaque_identity` around the write.) This might also end-up being relied-upon in practice. And an updated formal memory model will likely do 99% of the work of justifying such an idiom, which could be an interesting development for leveraging dependency ordering in other languages as well. (I initially wanted to suggest an improvement to the manual based on this, but by thinking more about it I realized that it requires some `opaque_identity` somewhere, so I just leave this for you as a comment.)

kayceesrk commented:
@gadmm Thanks for the insightful comments as always. 

> Concretely, after inlining, I wonder if some optimization (e.g. RL in the PLDI paper) could have d2 staying forever with the initial value None as a result of the load.

The memory model doesn't guarantee that `Some ..` will be seen. It would be surprising if the compiler eliminated the read in `d2`. The redundant load optimisation, for example, only optimises when it sees two loads. There is only one load of `r` in this program. 

>  One solution could be not to present it as a message-passing idiom but only as a memory-safety property.

Sounds like a good plan. I can present this as a "buggy" message passing example, where a non-atomic reference `r` is used whereas the correct implementation should have had `r` as an atomic reference. We then argue that memory safety is still preserved. 

> I believe that the text should make a clearer distinction between the intended practical memory model for OCaml and what has been presented in the PLDI paper, indeed the reader might incorrectly infer that an argument for memory safety follows from the referenced paper.

It is very likely that the person who is reading the memory model manual page is likely not to have read the PLDI paper. In that case, what would be the wording we should include such that it doesn't confuse the reader? I'd like to avoid pointing to some detail in the paper, that the reader likely hasn't read, and unintentionally sounding like that detail is important. As with any published paper, the paper is a snapshot of an idea in time, and it is likely that the idea itself evolves whereas the paper doesn't. I used the [Java](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html) and [c++](https://en.cppreference.com/w/c/atomic/memory_order) memory model documentation as guides for writing this page. One nice aspect of that document is that it aims to be self-contained. 

Could you suggest the text that should go in here?

kayceesrk commented:
@gadmm 

I've redone the "LDRF > memory model" section to be on memory safety. The ones to be addressed are the additional clarifications on (a) dependency ordering and (b) the diff between the PLDI paper and this manual chapter. Waiting for your text suggestions for both.

tomjridge commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,788 @@
+reference cell. One would expect that the assertion in the main function will
+never fail. The reasoning is that if "r2" is "0", then the write in "d2"
+occurred before the read of "b" in "d1". Given that "a" and "b" are aliases,
+the read of "a" in "d1" should also return "0".
```
Maybe say: the *second* read of "a" in "d1" ? (I think the first read can still return 2.)

kayceesrk replied:
Fixed.

tomjridge commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,788 @@
+the sequential meaning of the program. However, CSE breaks sequential
+reasoning.
+
+In the optimized program above, even if the write to be "b" in "d2" occurs
```
typo: write to *be* "b"

kayceesrk replied:
Fixed.

gadmm commented:
> @gadmm Thanks for the insightful comments as always.
> 
> > Concretely, after inlining, I wonder if some optimization (e.g. RL in the PLDI paper) could have d2 staying forever with the initial value None as a result of the load.
> 
> The memory model doesn't guarantee that `Some ..` will be seen. It would be surprising if the compiler eliminated the read in `d2`. The redundant load optimisation, for example, only optimises when it sees two loads. There is only one load of `r` in this program.

To clarify, I was speaking about the more general pattern where the user might be reading `r` several times. In fact I am convinced by the argument that the program must also poll often-enough and this forces `r` to be reloaded often-enough (at least until the compiler becomes smart-enough to know that `r` is not accessed from any asynchronous callback). So I liked the presentation of the data-passing idiom. But well this does not follow from the memory model.

> 
> > One solution could be not to present it as a message-passing idiom but only as a memory-safety property.
> 
> Sounds like a good plan. I can present this as a "buggy" message passing example, where a non-atomic reference `r` is used whereas the correct implementation should have had `r` as an atomic reference. We then argue that memory safety is still preserved.

I have read the new text and I like it.

> 
> > I believe that the text should make a clearer distinction between the intended practical memory model for OCaml and what has been presented in the PLDI paper, indeed the reader might incorrectly infer that an argument for memory safety follows from the referenced paper.
> 
> It is very likely that the person who is reading the memory model manual page is likely not to have read the PLDI paper. In that case, what would be the wording we should include such that it doesn't confuse the reader? I'd like to avoid pointing to some detail in the paper, that the reader likely hasn't read, and unintentionally sounding like that detail is important. As with any published paper, the paper is a snapshot of an idea in time, and it is likely that the idea itself evolves whereas the paper doesn't. I used the [Java](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html) and [c++](https://en.cppreference.com/w/c/atomic/memory_order) memory model documentation as guides for writing this page. One nice aspect of that document is that it aims to be self-contained.
> 
> Could you suggest the text that should go in here?

I think this is a good plan, and indeed going into low-level details such as dependency ordering at this particular place goes against the purpose of presenting a higher-level model. A better place could be at the start, when you cite the PLDI paper. You could clarify that the paper is augmented with details about publication-safety in the more recent GC, ideally citing alongside whatever outcome of https://github.com/ocaml/ocaml/pull/10995; this is not yet available but my hope is that it will result in some kind of even more expert-level documentation as an `.md` file in the tree. There was no progress recently on #10995 but it should be easy to copy-paste Stephen's various explanations into a file. Otherwise a quick explanation at the start of the manual.

kayceesrk commented:
Let me try to work in a sentence at the beginning about the difference between memory model in the paper and this document.

jhjourdan commented:
FTR, I have reviewed this off-line during my flight between California and Paris, I need a bit more time to actually send my comments. Sorry for the delay.

kayceesrk commented:
No worries @jhjourdan. Happy to wait for your comments.

jhjourdan commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,732 @@
+
+\begin{caml_example*}{verbatim}
+(* fib.ml *)
+let n = try int_of_string Sys.argv.(1) with _ -> 1
```
`try ... with _ -> ...` is usually considered as bad practice in OCaml. Perhpas we should not use it in the reference manual.

gasche replied:
we have `int_of_string_opt` now (since 4.05, according to `man stdlib`; time flies...)

damiendoligez replied:
The problem is not just `int_of_string` but also `Sys.argv.(1)`, which may raise `Invalid_argument`. I would suggest simply dropping the `try`...`with` as this code doesn't need to be bulletproof.

kayceesrk replied:
I've dropped the `try ... with`

jhjourdan commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,732 @@
+
+\begin{caml_example*}{verbatim}
+(* fib_twice.ml *)
+let n = try int_of_string Sys.argv.(1) with _ -> 1
```
Same remark as above.

kayceesrk replied:
"Fixed" as above.

jhjourdan commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,732 @@
+    2.85 ± 0.05 times faster than './spectralnorm.exe 4096'
+\end{verbatim}
+
+On the author's machine, the program scales well up to 4 domains but performs
+worse with 8 domains. This is due to the fact that unlike the parallel
+Fibonacci program, this program performs floating-point computations. At 8
+domains, the hyper-threaded siblings contend for the shared floating-point
+units on a core leading to worse behaviour than at 4 domains.
```
I'm not convinced by this analysis, because even for 2 and 4 domains the speedup is not so good (1.83 and 2.16, respectively).  My explanation is rather that the tasks are smaller, so that the synchronisation is more expensive on the second example.

kayceesrk replied:
>  My explanation is rather that the tasks are smaller,

`parallel_for` uses divide-and-conquer to distribute the work: https://github.com/ocaml-multicore/domainslib/pull/16. By default, each task should have enough chunked work to do. I would think that the tasks being small is not the issue. 

I explored this a bit more. I thought the issue was due to the impact of minor GCs. Compared to the parallel fibonacci example, this example allocates much more in the minor heap. 

```bash
kc@KCs-MacBook-Pro temp % OCAMLRUNPARAM="v=0x400" ./spectralnorm_par.exe 8 8192
1.274224153
allocated_words: 5369082705
minor_words: 5368902430
promoted_words: 96098
major_words: 276373
minor_collections: 25720
major_collections: 406
forced_major_collections: 0
heap_words: 249865
top_heap_words: 69641
mean_space_overhead: 867.869048
kc@KCs-MacBook-Pro temp % OCAMLRUNPARAM="v=0x400" ./fib_par2.exe 8 42
fib(42) = 433494437
allocated_words: 6864229
minor_words: 6864196
promoted_words: 11215
major_words: 11248
minor_collections: 54
major_collections: 3
forced_major_collections: 0
heap_words: 40960
top_heap_words: 40960
mean_space_overhead: 13567.659682
```

25k minor GCs in `spectralnorm_par` vs 54 in `fib_par2`. However, when I increased the size of the minor heap, while brought down the number of minor GCs, the running time remained the same.

In order to make progress on this manual page, I'm going to edit the paragraph as follows:

```
On the author's machine, the program scales reasonably well up to 4 domains but performs 
worse with 8  domains. Recall that the machine only has 4 physical cores. Debugging and 
fixing this performance issue is beyond the scope of this tutorial.
```

jhjourdan replied:
LGTM

jhjourdan commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,732 @@
+
+\section{s:par_atomics}{Atomics}
+
+Mutex, condition variables and semaphores are used to implement blocking
+synchronisation between domains. For non-blocking synchronisation, OCaml
+provides \stdmoduleref{Atomic} variables. As the name suggests, non-blocking
+synchronisation does not provide mechanisms for suspending and waking up
+domains. On the other hand, primitives used in non-blocking synchronisation are
+often compiled to atomic read-modify-write primitives that the hardware
+provides. As an example, the following program increments a non-atomic counter
+and an atomic counter in parallel.
```
I don't quite like this distinction between blocking and non-blocking synchronization:

1- Mutexes have non-blocking primitives, such as `try_lock`
2- Atomic variables may very well be used for purposes that are out of the scope of synchronization. We can expect that their SC behaviour be their main interesting feature for some usecases.

Do we really need to present atomic variables as a "synchronization primitive", rather than simply as a low-level mean of communication between threads?

kayceesrk replied:
Classification is hard :-) I'm using "synchronization" for atomic variables in the memory model sense. When a thread writes to a shared atomic reference which is read by another thread, the access is said to be synchronized.

jhjourdan commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,732 @@
+let twice_in_parallel f =
+  let d1 = Domain.spawn f in
+  let d2 = Domain.spawn f in
+  ignore @@ Domain.join d1;
+  ignore @@ Domain.join d2
```
```suggestion
  Domain.join d1;
  Domain.join d2
```

`ignore` is useless here because `Domain.join` returns `unit`.

kayceesrk replied:
Done.

jhjourdan commented on manual/src/tutorials/parallelism.etex:
```diff
@@ -0,0 +1,732 @@
+counter "curr" and update it to "curr + 1". Hence, instead of two increments,
+the effect will be that of a single increment. On the other hand, the atomic
+counter performs the load and the store atomically with the help of hardware
+support for atomicity. The atomic counter returns the expected result.
```
We could add here an additional explanation here to emphasize that this is expected because this piece of code contains a data race.

We could perhaps even note that because of the weak behavior of the memory model, the final value can be as low as 1. This is intersting, because this is a simple, concrete example where the memory model has an important impact.

kayceesrk replied:
> We could add here an additional explanation here to emphasize that this is expected because this piece of code contains a data race.

This is what I had originally :-) And @fpottier rightly pointed out that there is the simpler issue of non-atomicity. I'm in two minds about discussing the memory model issues here. The readers may not have sufficient understanding of the relaxed memory model issues to appreciate the observations that we may make.

kayceesrk replied:
I'm choosing to leave the current text as is here.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+
+\subsection{ss:mm_hw_opt}{Hardware optimisations}
+
+The other source of reordering is by the hardware. Modern hardware
+architectures have complex cache hierarchies with multiple levels of cache.
+While cache coherence ensures that reads and writes to a single memory
+location respect sequential consistency, the guarantees on programs that
+operate on different memory locations are much weaker. Consider the following
```
I heard from people understanding CPU architecture much better than me that cache was in fact not causing weak behaviours, because the cache coherence protocols ensured proper locking of memory that imply a global total order of all memory locations. Apparently, the sources of weak behaviours are rather:
- The store buffer
- Hardware speculation and rescheduling
- Compiler optimization

damiendoligez replied:
Operations may be reordered because one is fulfilled from a lower-level (and faster) cache than the other, even in the presence of cache-coherence protocols.

The store buffer is mentioned in the text after the example code.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+
+For a given execution trace, we define an irreflexive, transitive
+\emph{happens-before relation} that captures the causality between actions in
+the OCaml program. The happens-before relation is defined as follows:
```
```suggestion
the OCaml program. The happens-before relation is defined as the smallest transitive relation satisfying the following properties:
```

kayceesrk replied:
Done.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+  \item We define the order in which a domain executes its actions as the
+    \emph{program order}. If an action "x" precedes another action "y" in
+    program order, then "x" precedes "y" in happens-before order.
+	\item If "x" is an initialising write to a memory location (atomic or
+		otherwise) and "y" is any other read or write to that memory location, then
+		"x" precedes "y" in happens-before order.
```
Does this actually creates a full HB relationship? That seems really too strong to me. This essentially means that one can perform synchronisation without using any Atomic access or synchronisation library.

Consider this program:

```
let r = ref (ref 0)
let x = ref 0

let thread_a () =
    x := 1;
    r := ref 1

let thread_b () =
    let a = 2*!x in
    while !!r = 0 do () done;
    let b = 2*!x in
    (a, b)

let _ =
    Domain.create thread_b ();
    thread_a ();
    let (a, b) = Domain.join ();
    Printf.printf "%d %d" a b
```

If the above is correct, then there will be a HB relationship between the second read of x in thread_b and the write of x in thread_a. So the possible outcomes in the proposed memory model are (0, 2) and (2, 2). But since we want to allow CSE between the definitions of a and b, we want to also allow (0, 0).

To me, the solution to fix this is simple: remove these buggy HB relationships, and exclude initialisations from conflicting accesses. As far as I can tell the DRF-SC theorem still holds.

kayceesrk replied:
I have some thoughts on this point. I've pinged @stedolan to discuss this on a different channel to discuss this point.

kayceesrk replied:
Had a chat with @stedolan on this issue and he agrees that it is incorrect to have added initialising write to the HB relationship. I'll remove initialising writes from HB and change the examples appropriately.

kayceesrk replied:
Removed.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+reasoning. In this section, we will use the functions with names "dN" to
+represent domains executing in parallel with other domains.
+
+Here is a simple example with a data race:
```
Perhaps tell somewhere (once and for all) that d1 and d2 are supposed to be two different domains? I.e., tell that there is somewhere the code `let () = Domain.create d1 (); d2 ()`

kayceesrk replied:
Done.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+\end{caml_example*}
+
+Observe that "c" is a newly allocated reference. Can the read of "c" return a
+value which is not 42? Surprisingly, in the C++ and Java memory models, the
+answer is yes. With the C++ memory model, if the program has a data race, even
```
Well, this is not true in Java if `snippet` does not expose `c` to the rest of the world. So we should make `snippet` return `c` as well.

kayceesrk replied:
~How do you want this to be done? Could you suggest the edit here?~

I got what you meant. Changing `snippet` to return `c`. Fixed.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+...
+\end{verbatim}
+
+The OCaml memory model is bound both in space and time. The only memory location
+here is "c". Reasoning only about this snippet, there is neither the data race
```
What about `g`? It would be much clearer if `snippet` was really a part of the larger piece of code. This is made possible if `snippet` returns `c` as proposed above.

kayceesrk replied:
Done.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+\end{caml_example*}
+
+Observe that if "d2" does not witness the initialising writes on "s" or "o",
+the program may crash. LDRF allows us to reason about the program behaviour for
+a subset of memory locations. Let us ignore "r" and only consider "s" and "o".
+The objects "s" and "o" have initialising writes in "d1" and reads in "d2". The
+happens-before definition (section \ref{s:happens_before}) says that the
+initial write to a location precedes any other actions on that location. Hence,
+there is a happens-before relationship between the initial writes to "s" and
+"o" in "d1" and the reads in "d2". Hence, there is no data race on "s" and "o",
+and the program fragment that uses "s" and "o" will have sequentially
+consistent behaviour.
```
Of course, this has to be rewritten in light of my comment above.

kayceesrk replied:
I've removed this example as it relied on the reasoning through happens-before. The memory safety still holds but not due to happens-before reasoning.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+  \item Operations on mutexes.
+\end{itemize}
+
+\emph{(TODO: Include semaphores and condition variables in the inter-domain
+actions.)}
```
TODO, indeed.

Octachron replied:
This should be converted to a latex comment for now.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+to LDRF-SC, we can use sequential reasoning, which guarantees that "d2" will
+always see the initialising writes from "d1".
+
+\section{s:mm_semantics}{An operational view of the memory model}
```
Perhaps we could refer here to the PLDI paper?

kayceesrk replied:
Done.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+In this section, we describe the semantics of the OCaml memory model. Given an
+OCaml program, which may possibly contain data races, the semantics tells you
+the values that may be observed by the read of a memory location. For
+simplicity, we restrict the intra-thread actions to just the accesses to atomic
+and non-atomic locations, ignoring domain spawn and join operations, and the
+operations on mutexes.
```
Why so? Does it really complicate the model so much?

kayceesrk replied:
Yes, I would think so. We'll have to model the fact that a domain may block on a mutex or domain join. While this shouldn't be too hard, the additional details may make it harder to appreciate the semantics of the memory access, which is the focus in this section. 

At the very least, any extension to the basic presentation will need to be reviewed very closely. The operational semantics presented here closely follows the PLDI 2018 operational semantics.

jhjourdan replied:
I don't have a particularly strong opinion on this and I don't want to block anything, but I indeed think that the PLDI 2018 presentation is lacking some important aspects. I think it would be a good think to include such a more in-depth description here because:
- As experience shows (e.g., allocations), the memory model for the full language is never a "trivial extension", and subtle problems can arise unexpectedly (e.g., how one should formulate the model to make sure that a read is always safe even though it is not happens-after the allocation of the cell...). I understand that this text is targeted at OCaml developpers which are not necessarily interested in all these details, but if it's not described in the OCaml manual, then where else should it be described?
- I don't think adding explanations for spawn/join/allocations/mutexes/... is invasive in the sense that it would require changing the text for the other operations. The general setting (frontiers...) need not be changed.

kayceesrk replied:
> As experience shows (e.g., allocations), the memory model for the full language is never a "trivial extension"

Indeed. This matched our experience of implementing the model. I have a sufficient understanding of how allocation and initialisation work in the memory model now. 

Perhaps the right approach would be to first write a peer-reviewed paper that covers allocations and initialisations, compilation to LKMM and reasoning about concurrent accesses to shared locations from OCaml (hardware model) and C (LKMM). And then we can aim to summarize that in the manual. I am hesitant to include new tricky material on the memory model directly in the manual. What do you think?

Could you point me to memory model papers that explicitly deal with allocations?

> I don't think adding explanations for spawn/join/allocations/mutexes/... is invasive in the sense that it would require changing the text for the other operations. The general setting (frontiers...) need not be changed.

In a follow-up PR, I'll try to extend the operational semantics with spawn, join, mutexes, condition and semaphore.

jhjourdan replied:
> Perhaps the right approach would be to first write a peer-reviewed paper that covers allocations and initialisations, compilation to LKMM and reasoning about concurrent accesses to shared locations from OCaml (hardware model) and C (LKMM). And then we can aim to summarize that in the manual. I am hesitant to include new tricky material on the memory model directly in the manual. What do you think?

Seems good to me. The memory model of OCaml and its interaction with C are two different matters.

> Could you point me to memory model papers that explicitly deal with allocations?

Our work on Cosmo deals with allocations.

> In a follow-up PR, I'll try to extend the operational semantics with spawn, join, mutexes, condition and semaphore.

:+1:

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+"d1" performs the read of "b". For non-atomic reads, the domains may read an
+arbitrary element of the history for that location, as long as it is not older
+than the timestamp in the domains's frontier. In this case, since "d1" frontier
+at "b" is at "t3", the read may return the value "3", "4" or "5".
```
Add a comment saying that a non-atomic read does not change the frontier of the current domain. This is far from obvious.

kayceesrk replied:
Done.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+\end{verbatim}
+
+As a result of the atomic read, the abstract machine state is updated to:
+
+\begin{verbatim}
+(* Non-atomic locations *)
+a: [t1 -> 1; t2 -> 2]
+b: [t3 -> 3; t4 -> 4; t5 -> 5]
+c: [t6 -> 5; t7 -> 6; t9 -> 10; t8 -> 7]
+
+(* Domains *)
+d1: [a -> t2; b -> t4; c -> t6] (* frontier updated at a and b *)
+d2: [a -> t1; b -> t4; c -> t9]
+
+(* Atomic locations *)
+A: 10, [a -> t1; b -> t5; c -> t7]
+B: 5,  [a -> t2; b -> t4; c -> t6]
+\end{verbatim}
```
The example is not really well-chosen because the frontier of B is <= that of d1, so that we don't see that d1 keeps the timestamps of locations it knows better than B.

kayceesrk replied:
Did you mean to say that before the read, for all the locations, frontier of d1 is <= the frontier of B? For reference the frontiers before the read are:

```
d1: [a -> t1; b -> t3; c -> t6]
B : [a -> t2; b -> t4; c -> t6]
```

I'll change the frontier of `d1@c` to `t7`. This should address the concern that `d1` may know more recent writes than `B`.

kayceesrk replied:
Done.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+  \item The "Bytes" module~\stdmoduleref{Bytes} permits mixed-mode accesses
+    where reads and writes may be of different sizes. Unsynchronized mixed-mode
+    accesses lead to tearing.
+\end{itemize}
```
What about flat float mutable records?

damiendoligez replied:
For float mutable records, all reads and writes are the same size, right?

damiendoligez replied:
Oh, you mean flat float records on 32-bit architectures.

kayceesrk replied:
Indeed. Can you either provide me the suggested text or feel free to push the change directly to the branch?

jhjourdan replied:
"With flat-float arrays on 32-bit architectures, getting or setting a field involves two separate memory accesses. In the presence of data races, the user may observe tearing."

becomes:

"With flat-float arrays or records with only float fields on 32-bit architectures, getting or setting a field involves two separate memory accesses. In the presence of data races, the user may observe tearing."

kayceesrk replied:
Done.

jhjourdan reviewed (changes requested):
Finally, here is my review. Apart from several minor concerns, I have a major concern about the HB relationship which is defined in the document: I don't think it is a good thing to include a HB relation between an initialization and any accesses to the same memory location.

gadmm commented:
@kayceesrk (and perhaps this could interest @lthls also) I was wondering whether you had considered specifying the way asynchronous callbacks (finalizers, signal handlers, memprof callbacks) communicate with their main thread, by saying that they behave as if they are on different threads as far as read/writes to mutable locations are concerned.

More precisely I was wondering whether there could be a path towards the compiler (e.g. flambda) not having to consider that any mutable field could be read or written to at every safe point, in order to make code optimisations apply more often, in the long run. Concretely if you have two loads `!x` and `!x` across an allocation, then you should be able to do CSE. IIRC Vincent's explanations, Flambda will never do such CSE because it assumes that arbitrary code can run at any point due to async callbacks.

Due to backwards-compatibility, this is best specified early-on after introducing the memory model. Also the patterns that I know in existing code would not break due to such a change (concretely nobody does `while !x do poll_async() done`).

lthls commented:
> More precisely I was wondering whether there could be a path towards the compiler (e.g. flambda) not having to consider that any mutable field could be read or written to at every safe point, in order to make code optimisations apply more often, in the long run. Concretely if you have two loads `!x` and `!x` across an allocation, then you should be able to do CSE. IIRC Vincent's explanations, Flambda will never do such CSE because it assumes that arbitrary code can run at any point due to async callbacks.

It's not really an issue for Flambda, as we're not interested in mutable structures anyway. However, you can look at the following lines in `CSEgen.ml`: https://github.com/ocaml/ocaml/blob/1d65f5b55922e5649ce74e2143c9cbbcd97d0163/asmcomp/CSEgen.ml#L290-L294
If the memory model states that asynchronous code behaves as if it was executing in another thread, then you should be able to remove the call to `kill_loads`.

OlivierNicole commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,790 @@
+  \item With flat-float arrays on 32-bit architectures, getting or setting a
+    field involves two separate memory accesses. In the presence of data races,
+    the user may observe tearing.
+  \item The "Bytes" module~\stdmoduleref{Bytes} permits mixed-mode accesses
+    where reads and writes may be of different sizes. Unsynchronized mixed-mode
+    accesses lead to tearing.
```
Maybe we should document the fact that some `Bytes` functions step out of the memory model. E.g.:
```
    \item Unsynchronized accesses to the same sub-range of "bytes" using
      functions from the \stdmoduleref{Bytes} module, such as "Bytes.blit" or
      "Bytes.fill", are an exception to the memory model. Concurrent writes or
      concurrent reads and writes to the same bytes invalidate any guarantees
      about the program. This warning extends to all standard functions that
      writes to "bytes" directly, such as "Unix.recv".
```

jhjourdan replied:
First, the problem is about accesses to bytes of the same byte array, not on ly to the same bytes. We could even say that there is a problem when accessing to bytes in the same 64bits (or 32 bits) block in a byte array only.

Second, we should take care about bounding the undefinedness to the array of bytes. This is the general spirit of the memory model that a race condition is bounded in space and time. I.e., we could say that all bets are off when it comes to values read from that bytes array until the next synchronized write.

OlivierNicole replied:
Regarding your second point, it seems that we have a choice between accepting a possibility of C-level undefined behavior with this function, not bounded in space and time, or making it much slower by performing a byte-per-byte copy.

jhjourdan replied:
Or we could use in C relaxed atomic accesses that do not suffer from UB, but do not incur significant overheads since they are compiled to plain accesses.

I think that breaking type safety because of this is not an option.

kayceesrk commented:
@jhjourdan I've addressed your major comment about the initialising writes being part of the happens-before. I have also addressed the minor comments. Would you be able to review this again please? 

At some point, we will need to redo the memory model paper fixing this one bug, but also extending the model to include the 

1. compilation of the memory model to the Linux kernel memory model (LKMM) that we use in our C FFI / runtime. 
2. semantics with initialising writes and their compilation to hardware memory models and LKMM. 

p.s: As I was addressing the comments, I've marked as "resolved" those comments that I think need not be reviewed again. If you think this workflow makes reviewing hard, let me know and I shall unresolve all the resolved comments.

jhjourdan commented on manual/src/tutorials/memorymodel.etex:
```diff
@@ -0,0 +1,756 @@
+\subsection{s:datarace}{Data race}
+
+In a given trace, two actions are said to be \emph{conflicting} if they access
+the same non-atomic location, and at least one is a write.
```
There should not be a conflict between an initializing write and any further access (read or write) to the same memory location (these are not races). These should be removed from the definition of conflicting accesses.

kayceesrk replied:
Good catch. I've changed the definition to:

```
In a given trace, two actions are said to be \emph{conflicting} if they access   
the same non-atomic location, at least one is a write and neither is an               
initialising write to that location.
```

jhjourdan commented:
@kayceesrk I had a brief look at the current state and the changes you made. This seems good to me, except the few minor comments I made.

kayceesrk committed:
Add Changes entry

kayceesrk commented:
Thanks everyone for the comments. I've squashed the commit history and added a Changes entry. This PR is ready to merge. 

As future work, we need the following:

1. Include semaphores and condition in the inter-domain actions https://github.com/ocaml/ocaml/pull/11280#discussion_r913019398
2. Extend the operational semantics to include allocations, mutex, condition and semaphores https://github.com/ocaml/ocaml/pull/11280#discussion_r981851959
3. Describe the interaction with C FFI (that C FFI preserves local data-race freedom)

Octachron committed:
manual: put TODO in comments

Octachron commented:
I have pushed a commit that transforms the remaining TODO comments into a latex comment (with @kayceesrk approval). I will squash and merge the PR once the CI returns greens.
